File: lib_audio/src/main/java/com/imooc/lib_audio/mediaplayer/core/CustomMediaPlayer.java
package com.imooc.lib_audio.mediaplayer.core;

import android.media.MediaPlayer;

import java.io.IOException;

public class CustomMediaPlayer extends MediaPlayer implements MediaPlayer.OnCompletionListener {

	private Status mState = Status.IDLE;
	private OnCompletionListener mListener;

	public enum Status{
		IDLE, INITIALIZED, STARTED, PAUSED, STOPPED, COMPLETED, PREPARED
	}

	CustomMediaPlayer(){
		super();
		mState = Status.IDLE;
		super.setOnCompletionListener(this);
	}

	@Override
	public void onCompletion(MediaPlayer mp) {
		mState = Status.COMPLETED;
		if(mListener != null){
			mListener.onCompletion(mp);
		}
	}

	@Override
	public void setDataSource(String path) throws IOException, IllegalArgumentException, IllegalStateException, SecurityException {
		super.setDataSource(path);
		mState = Status.INITIALIZED;
	}

	@Override
	public void start() throws IllegalStateException {
		super.start();
		mState = Status.STARTED;
	}


	@Override
	public void pause() throws IllegalStateException {
		super.pause();
		mState = Status.PAUSED;
	}

	@Override
	public void stop() throws IllegalStateException {
		super.stop();
		mState = Status.STOPPED;
	}

	@Override
	public void reset() {
		super.reset();
		mState = Status.IDLE;
	}

	@Override
	public void prepare() throws IOException, IllegalStateException {
		super.prepare();
		mState = Status.PREPARED;
	}

	public void setState(Status mState) {
		this.mState = mState;
	}

	public Status getState() {
		return mState;
	}

	public boolean isComplete() {
		return mState == Status.COMPLETED;
	}

	public void setOnCompletionListener(OnCompletionListener listener) {
		this.mListener = listener;
	}
}

File: lib_audio/src/main/java/com/imooc/lib_audio/mediaplayer/core/CustomMediaPlayer.kt
package com.imooc.lib_audio.mediaplayer.core

import android.media.MediaPlayer

class CustomMediaPlayer: MediaPlayer, MediaPlayer.OnCompletionListener{


    private var mStatus : Status = Status.IDLE
    private lateinit var mListener : OnCompletionListener

    constructor() : super() {
        mStatus = Status.IDLE
        super.setOnCompletionListener(this)
    }

    enum class Status{
        IDLE, INITIALIZED, STARTED, PAUSED, STOPPED, COMPLETED, PREPARED
    }



    //播放完毕
    override fun onCompletion(mediaPlayer: MediaPlayer?) {
        mStatus = Status.COMPLETED
        mListener.onCompletion(mediaPlayer)
    }

    override fun setDataSource(path: String?) {
        super.setDataSource(path)
        mStatus = Status.INITIALIZED
    }

    override fun start() {
        super.start()
        mStatus =  Status.STARTED
    }

    override fun stop() {
        super.stop()
        mStatus = Status.STOPPED
    }

    override fun pause() {
        super.pause()
        mStatus = Status.PAUSED
    }

    override fun reset() {
        super.reset()
        mStatus = Status.IDLE
    }

    override fun prepare() {
        super.prepare()
        mStatus = Status.PREPARED
    }

    fun getState() : Status{
        return mStatus
    }

    override fun setOnCompletionListener(listener: OnCompletionListener?) {
        if (listener != null) {
            this.mListener = listener
        }
    }
}